### PR TITLE
Initial server stuff

### PR DESCRIPTION
**Pull request by LordAro**
May be broken. Not ready for merge.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -27,19 +27,36 @@ Options:
...
-	args, err := docopt.Parse(usage, nil, true, "ury-listd-go 0.0", false)
+	args, err := parseArgs()
 	if err != nil {
 		logger.Fatal("Error parsing args: " + err.Error())
 	}
+
+	// Set up server listener
+	serverCh := make(chan string)
+	server, err := MakeServer(args["--addr"].(string), args["--port"].(string), serverCh)
+	if err != nil || server == nil {
```
Comment (wlcx) [inline main.go]: Surely if err is not nil server won't be? And even if err is nil but server is not you'll print nil?

Comment (wlcx) [inline main.go]: Nowait, you'll cause a panic for calling nil.Error()

Comment (LordAro) [inline main.go]: fixed locally

**Comment by MattWindsor91** [comment]
Let us know when this can be merged :heart:

File: server.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+type Server struct {
+	listener   net.Listener
+	serverCh   chan baps3.Message
+	clientComm chan []byte
+	tok        *baps3.Tokeniser
+	clients    []*Client
+}
+
+func MakeServer(addr string, port string, serverCh chan baps3.Message) (*Server, error) {
```
Comment (wlcx) [inline server.go]: I personally prefer NewServer

**Comment by wlcx** [comment]
So rename Server->Listener as discussed? :)

File: main.go [hunk 3]
```diff
@@ -27,29 +27,47 @@ Options:
...
-	connector := baps3.InitConnector("", responseCh, wg, logger)
+	connLog := log.New(os.Stderr, "playd: ", 0)
+	connector := baps3.InitConnector("", responseCh, wg, connLog)
 	connector.Connect(args["--playoutaddr"].(string) + ":" + args["--playoutport"].(string))
 	go connector.Run()
+
+	// Set up server listener (requires connector.ReqCh)
+	listener, err := MakeListener(args["--addr"].(string), args["--port"].(string), connector.ReqCh)
+	if err != nil {
+		log.Fatal("Error initialising connection server: " + err.Error())
```
Comment (wlcx) [inline main.go]: s/server/listener

File: listener.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	"log"
+	"net"
+
+	baps3 "github.com/UniversityRadioYork/baps3-go"
+)
+
+type Listener struct {
+	listener    net.Listener
+	connectorCh chan baps3.Message
+	clientComm  chan []byte
```
Comment (wlcx) [inline listener.go]: Would these two channels be better names responseCh and requestCh as in baps3.Connector?

File: listener.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		clientComm:  make(chan []byte),
+		connectorCh: connectorCh,
+	}
+	return listener, nil
+}
+
+func (l *Listener) Broadcast(msg baps3.Message) {
+	// Probably should have \n at the end
+	for _, c := range l.clients {
+		data, err := msg.Pack()
```
Comment (wlcx) [inline listener.go]: surely pack the message once and then send it to all clients?

File: client.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+func MakeClient(conn net.Conn, clientCh chan []byte) *Client {
+	client := &Client{
+		conn:       conn,
+		reader:     bufio.NewReader(conn),
+		writer:     bufio.NewWriter(conn),
+		listenerCh: clientCh,
+		Outgoing:   make(chan []byte),
+	}
+
+	go client.Read()
```
Comment (wlcx) [inline client.go]: This currently never terminates, unless I'm missing something obvious. There should be a way to gracefully kill the goroutines for a connection

Comment (wlcx) [inline client.go]: Yeah, as soon as a client disconnects listd maxes out a core

File: listener.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+		if err != nil {
+			continue // TODO Handle
+		}
+		c.Outgoing <- data
+	}
+}
+
+func (l *Listener) ReceiveLoop() {
+	for data := range l.clientComm {
+		lines, _, err := l.tok.Tokenise(data)
```
Comment (wlcx) [inline listener.go]: Clients should not share a tokeniser really, as multiple partial requests will interfere with each other. This probably doesn't show itself as the client's reader blocks until `\n` but this is still not ideal. Every client should have it's own tokeniser and only send complete baps3.Message structs to the listener.

File: client.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+func (c *Client) Read() {
+	for {
+		data, _ := c.reader.ReadBytes('\n')
+		c.listenerCh <- data // Each client doesn't care what it got, that's for the server to handle
+	}
+}
+
+func (c *Client) write() {
```
Comment (wlcx) [inline client.go]: So, way to exit this goroutine is to close(c.Outgoing), at which point the for loop will exit

File: client.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	for data := range c.Outgoing {
+		_, err := c.writer.Write(data)
```
Comment (wlcx) [inline client.go]: I don't believe there is any benefit using a bufio.Writer here, as we have complete packed messages ready to go at once. Just call c.conn.Write(data) directly.

File: client.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	return baps3.NewMessage(baps3.RsOhai).AddArg("listd").AddArg("0.0")
+}
+
+func (c *Client) Close() {
+	c.conn.Close()
+}
+
+func (c *Client) Read() {
+	for {
+		data, _ := c.reader.ReadBytes('\n')
```
Comment (wlcx) [inline client.go]: You should check err here, as it should be != nil if the connection has be closed the other end.

File: listener.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+		}
+	}
+}
+
+// Writes new responses to the client connection.
+// New responses are got from resCh. Errors in writing the data
+// will cause the connection to be disconnected, via rmCh.
+func (c *Client) Write(resCh <-chan baps3.Message, rmCh chan<- *Client) {
+	for {
+		msg, more := <-resCh
```
Comment (wlcx) [inline listener.go]: would suggest s/more/ok but a very small nitpick

File: listener.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+		if err != nil {
+			log.Println("Error writing from", c.conn.RemoteAddr(), ":", err.Error())
+			rmCh <- c
+			return
+		}
+	}
+}
+
+func makeWelcomeMsg() *baps3.Message {
+	return baps3.NewMessage(baps3.RsOhai).AddArg("listd").AddArg("0.0")
```
Comment (wlcx) [inline listener.go]: Might be worth having a global version variable for future compile-time injection? Also used in docopts iirc

File: listener.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+func makeFeaturesMsg() *baps3.Message {
+	return baps3.NewMessage(baps3.RsFeatures).AddArg("lol")
```
Comment (wlcx) [inline listener.go]: wat

File: listener.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+			close(client.resCh)
+			delete(clients, client.conn)
+			log.Println("Closed connection from", client.conn.RemoteAddr())
+		}
+	}
+}
+
+// Listens for new connections on addr:port and spins up the relevant goroutines.
+// cReqCh & cResCh are from the connector, requests get pushed down and responses get pulled, respectively.
+func runListener(addr string, port string, cReqCh chan<- baps3.Message, cResCh <-chan baps3.Message) {
```
Comment (wlcx) [inline listener.go]: runListener still has no way of exiting cleanly, AFAICT

File: listener.go (hunk 12 above)
Comment (LordAro) [inline listener.go]: yeah, that's on the todo list. just don't like having to remember to do -ldflags=... everytime i compile

File: listener.go (hunk 13 above)
Comment (LordAro) [inline listener.go]: well, there aren't any actual features yet...

File: listener.go (hunk 14 above)
Comment (LordAro) [inline listener.go]: ...does it need one?

File: listener.go [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+	clients := make(map[net.Conn]chan<- baps3.Message)
+
+	for {
+		select {
+		case msg := <-cResCh:
+			processResponse(&clients, msg)
+		case msg := <-reqCh:
+			processRequest(cReqCh, msg)
+		case client := <-addCh:
+			clients[client.conn] = client.resCh
```
Comment (wlcx) [inline listener.go]: Perhaps use a map[Client]bool and access the whole client object

Comment (LordAro) [inline listener.go]: Part of the point of this sort of setup is that i don't need to keep track of the Client objects, just the raw connections (and channels) - they keep track of themselves (woo garbage collection)

File: listener.go [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+	// Register user
+	h.addCh <- client
+
+	go client.Read(h.reqCh, h.rmCh)
+	client.Write(client.resCh, h.rmCh)
+}
+
+func makeWelcomeMsg() *baps3.Message {
+	return baps3.NewMessage(baps3.RsOhai).AddArg("listd").AddArg("0.0")
```
Comment (MattWindsor91) [inline listener.go]: According to [the spec](http://universityradioyork.github.io/baps3-spec/comms/internal/core.html#__code_ohai_code_server_welcome), this should only have _one_ argument, so something like 'listd-0.0'.  (It doesn't seem to say anything about how the version should be formatted, but this seems sensible.)

File: listener.go (hunk 13 above)
Comment (MattWindsor91) [inline listener.go]: Just send no arguments (`FEATURES` on its own) for now?

File: listener.go [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+	netListener, err := net.Listen("tcp", addr+":"+port)
+	if err != nil {
+		log.Println("Listening error:", err.Error())
+		return
+	}
+
+	go h.handleChannels()
+
+	// Get new connections
+	for {
```
Comment (wlcx) [inline listener.go]: IMO this should be an anonymous function gouroutine, and handlechannels should be the main body of this function, with an added case receive from a h.quit channel, through which the main goroutine can terminate it properly. There should also probably be a waitgroup somewhere.

**Comment by wlcx** [comment]
So I still maintain that listd needs a clean way to terminate. At the moment sigints /are/ handled, and the downstream connection is terminated gracefully, then listd dies, upstream connections be damned.
I would propose:
- Shutting down the listener accept goroutine like [this](http://zhen.org/blog/graceful-shutdown-of-go-net-dot-listeners/)
- Having a client.Close() method, which quits the connection gracefully, and call this on every client at quit time. Would also be a good reason to keep track of the whole client object but meh :shipit:

File: listener.go (hunk 16 above)
Comment (LordAro) [inline listener.go]: k
how do we cope with the playd ohai? then listd's ohai should be something like "playd-<hash>/listd-0.0" ?

**Comment by wlcx** [comment]
Close enough for now :p

**Pull request merged**